network/network_interface.ml

```diff
@@ -25,299 +25,609 @@ let uri () = "file:" ^ !default_path
-
-external set_gateway_interface: debug_info -> name:iface -> unit = ""
-external set_dns_interface: debug_info -> name:iface -> unit = ""
-
-module Interface = struct
-	external get_all : debug_info -> unit -> iface list = ""
-	external exists : debug_info -> name:iface -> bool = ""
-	external get_mac : debug_info -> name:iface -> string = ""
-	external is_up : debug_info -> name:iface -> bool = ""
-	external get_ipv4_addr : debug_info -> name:iface -> (Unix.inet_addr * int) list = ""
-	external set_ipv4_conf : debug_info -> name:iface -> conf:ipv4 -> unit = ""
-	external get_ipv4_gateway : debug_info -> name:iface -> Unix.inet_addr option = ""
-	external get_ipv6_addr : debug_info -> name:iface -> (Unix.inet_addr * int) list = ""
-	external get_dns : debug_info -> name:iface -> Unix.inet_addr list * string list = ""
-	external get_mtu : debug_info -> name:iface -> int = ""
-	external get_capabilities : debug_info -> name:iface -> string list = ""
-	external is_connected : debug_info -> name:iface -> bool = ""
-	external is_physical : debug_info -> name:iface -> bool = ""
-	external has_vlan: debug_info -> name:iface -> vlan:int -> bool = ""
-	external bring_down : debug_info -> name:iface -> unit = ""
-	external set_persistent : debug_info -> name:iface -> value:bool -> unit = ""
-	external make_config : debug_info -> ?conservative:bool -> config:(iface * interface_config_t) list-> unit -> unit = ""
+  let replace_assoc key new_value existing =
+    (key, new_value) :: (List.filter (fun (k, _) -> k <> key) existing) in
+
+  if List.mem_assoc name config then begin
+    replace_assoc name data config
+  end else
+    (name, data) :: config
+
+type errors =
+  | Script_missing of string
+        [@doc [
+          "[Script_missing (script)] is reported ";
+          "if unable to find [script]"]]
+  | Script_error of (string * string) list
+        [@doc [
+          "[Script_error ([(key * value); ...])] is reported ";
+          "when error occurs when executing script, the [key] and [value] indicates ";
+          "the infomation about the script and the error"]]
+  | Read_error of string
+        [@doc [
+          "[Read_error (file)] is reported ";
+          "when error occurs when reading [file]"]]
+  | Write_error of string
+        [@doc [
+          "[Write_error (file)] is reported ";
+          "when error occurs when writing [file]"]]
+  | Not_implemented
+      [@doc [
+        "[Not_implemented] is reported ";
+        "if the interface is not implemented"]]
+  | Vlan_in_use of (string * int)
+        [@doc [
+          "[Vlan_in_use (bridge, vlan_id)] is reported ";
+          "when [vlan_id] on [bridge] is inuse"]]
+  | PVS_proxy_connection_error
+      [@doc [
+        "[PVS_proxy_connection_error] is reported ";
+        "when unable to connect PVS proxy"]]
+  | Unknown_error
+      [@doc [
+        "The default variant for forward compatibility."
+      ]]
+[@@default Unknown_error]
+[@@deriving rpcty]
```

Comment:
These are fine, when you have such long types it make sense to append them to the bottom.

Not that you can replace the decorated strings with actual docstrings:
```ocaml
 [@doc [
        "[PVS_proxy_connection_error] is reported ";
        "when unable to connect PVS proxy"]]
```
into
```ocaml
 (** [PVS_proxy_connection_error] is reported
     "when unable to connect PVS proxy *)
```
and in this way they will appear both in the rpc documentation and in the ocaml one.

Reply:
That should be more readable, but where to put the docstring? before the match pattern or after?
e.g.
```
| PVS_proxy_connection_error 
(** [PVS_proxy_connection_error] is reported
     "when unable to connect PVS proxy *)
```
or 
```
| (** [PVS_proxy_connection_error] is reported
     "when unable to connect PVS proxy *)
PVS_proxy_connection_error 
```

Reply:
They go after, you can use https://caml.inria.fr/pub/docs/manual-ocaml/ocamldoc.html#sec349 as a reference

Reply:
Thank you!